torch/utils/data/datapipes/utils/decoder.py

```diff
@@ -0,0 +1,270 @@
+
+
+def torch_video(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in "mp4 ogv mjpeg avi mov h264 mpg webm wmv".split():
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchvision.io  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchvision.io.read_video(fname)
+
+
+################################################################
+# torchaudio
+################################################################
+
+
+def torch_audio(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in ["flac", "mp3", "sox", "wav", "m4a", "ogg", "wma"]:
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchaudio  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchaudio.load(fname)
+
+
+
+################################################################
+# a sample decoder
+################################################################
+
+
+class Decoder:
+    """
+    Decode key/data sets using a list of handlers.
+    For each key/data item, this iterates through the list of
+    handlers until some handler returns something other than None.
+    """
+
+    def __init__(self, handlers):
+        self.handlers = handlers
+
+    def add_handler(self, handler):
+        if not handler:
+            return
+        if not self.handlers:
+            self.handlers = [handler]
+        else:
+            self.handlers.append(handler)
+
+    def decode1(self, key, data):
+        if not data:
+            return data
+
+        # if data is a stream handle, we need to read all the content before decoding
+        if isinstance(data, io.BufferedIOBase) or isinstance(data, io.RawIOBase):
+            data = data.read()
+
+        for f in self.handlers:
+            result = f(key, data)
+            if result is not None:
+                return result
+        return data
+
+    def decode(self, datapipe):
+        result = {}
+        # single data tuple(pathname, data stream)
+        if isinstance(datapipe, tuple):
+            datapipe = [datapipe]
+
+        if datapipe is not None:
+            for k, v in datapipe:
+                # TODO: xinyu, figure out why Nvidia do this?
+                if k[0] == "_":
+                    if isinstance(v, bytes):
+                        v = v.decode("utf-8")
+                        result[k] = v
+                        continue
+                result[k] = self.decode1(k, v)
+        return result
+
+    def __call__(self, datapipe):
+        return self.decode(datapipe)
```

Comment:
Can we change `datapipe` to `data`, this name makes the logic confusing, as the input is actually one element for each iteration rather than the datapipe itself.
Besides, is there any benefit using dict as result?

Reply:
Will change datapipe to data. 
It is not clear whether dict has benefit at the moment, I am not sure whether we have any case that we need to decode a batch of image at once and need to locate a specific record within a dict. Let's keep this, we can change the code later to meet our needs.